sql/schemachanger: Block explicit transactions for ALTER COLUMN TYPE

spilchen opened the pull request:
This introduces a block that prevents running an ALTER COLUMN TYPE operation requiring a backfill within a transaction. To support the schemachanger end-to-end tests, which need to execute ALTERs in a transaction, I added a session variable to override this default behavior.

Epic: CRDB-25314
Informs #49329
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/136207)
<!-- Reviewable:end -->

rafiss reviewed (approved):
lgtm, just left an optional comment in case you think the suggestion makes sense

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136207#-:-OCd06KU531G0z9Afifl:b-89sk7q)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_type.go` line 437 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136207#-OCd-f37EUrwGOWIwv9K:-OCd-f37EUrwGOWIwv9L:bykmeyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/43be5d5c289fe1b682e75a9b7e3f510ef85eaffc/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_type.go#L437)):*
> ```Go
> // raises an error.
> func failIfExplicitTransaction(b BuildCtx) {
> 	if !b.SessionData().AllowAlterColumnTypeInExplicitTxn && !b.EvalCtx().TxnIsSingleStmt {
> ```

would it make sense to re-use the `UseNewSchemaChangerUnsafeAlways` session variable for this purpose?


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: Block explicit transactions for ALTER COLUMN TYPE

This introduces a block that prevents running an ALTER COLUMN TYPE
operation requiring a backfill within a transaction. To support the
schemachanger end-to-end tests, which need to execute ALTERs in a
transaction, I added a session variable to override this default
behavior.

Epic: CRDB-25314
Informs #49329
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136207)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_type.go` line 437 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136207#-OCd-f37EUrwGOWIwv9K:-OCdid5vCwEf40XMmU8B:b-mkfm86) ([raw file](https://github.com/cockroachdb/cockroach/blob/43be5d5c289fe1b682e75a9b7e3f510ef85eaffc/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_type.go#L437)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

would it make sense to re-use the `UseNewSchemaChangerUnsafeAlways` session variable for this purpose?
</blockquote></details>

That's a great idea—it avoids the need to introduce another variable.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
